### ISSUE TITLE
Add aerosol bypass in CRTM

### ISSUE THREAD

**Issue by chengdang** [JCSDA/CRTMv3#114]
In UFO, the `number of aerosols` of input profiles is set as the `number of aerosols in selected geovals`:
https://github.com/JCSDA-internal/ufo/blob/5e32c4ec5663a8350f7df123fb680e4c2ea9adc1/src/ufo/operators/crtm/ufo_crtm_utils_mod.F90#L1235

In practice, it is not necessarily ture that all aerosol will be assigned to a valid type in the optical LUT, subroutine https://github.com/JCSDA-internal/ufo/blob/5e32c4ec5663a8350f7df123fb680e4c2ea9adc1/src/ufo/operators/crtm/ufo_crtm_utils_mod.F90#L1255

The unassigned aerosol type `atm(m)%aerosol(i)%TYPE` will be set as the default value 0 during data allocation that triggers error report in CRTM, as noticed by @jeromebarre 

After discussion, it seems useful to introduce a type index as the bypass aerosol in CRTM. This allows UFO to more flexibility pass aerosol inputs to CRTM. The progress of implementation and testing will be documented in this issue.

**Comment by chengdang** [comment]
My initial thoughts was to create a function that remove all input aerosol profiles with type index of -1 as part of input data processing, see below:
```
  ELEMENTAL SUBROUTINE CRTM_Atmosphere_Aerosol_Bypass( Atmosphere )
    TYPE(CRTM_Atmosphere_type), INTENT(IN OUT) :: Atmosphere
    INTEGER,  ALLOCATABLE :: Aerosol_Type(:)
    REAL(fp), ALLOCATABLE :: Effective_Radius(:,:)
    REAL(fp), ALLOCATABLE :: Effective_Variance(:,:)
    REAL(fp), ALLOCATABLE :: Concentration(:,:)
    INTEGER :: n_aerosol_invalid, n_aerosol_valid
    INTEGER :: i, j
    INTEGER :: alloc_stat

    ! Do nothing if structure is unused
    IF ( .NOT. CRTM_Atmosphere_Associated(Atmosphere) ) RETURN

    ! Count the number of bypassed aerosols
    n_aerosol_invalid = 0
    DO i = 1, Atmosphere%n_Aerosols
      IF (Atmosphere%Aerosol(i)%Type == -1) THEN
          n_aerosol_invalid = n_aerosol_invalid + 1
      END IF
    END DO
    n_aerosol_valid = Atmosphere%n_Aerosols - n_aerosol_invalid

    ! Allocate aerosol related data structures
    ALLOCATE(Aerosol_Type( n_aerosol_valid ), &
             Effective_Radius( n_aerosol_valid, Atmosphere%n_Layers ), &
             Effective_Variance( n_aerosol_valid, Atmosphere%n_Layers  ), &
             Concentration( n_aerosol_valid, Atmosphere%n_Layers  ), &
             STAT = alloc_stat )
    IF ( alloc_stat /= 0 ) RETURN

    ! save valid aerosol types/concentrations
    j = 1
    DO i = 1, Atmosphere%n_Aerosols
        IF (Atmosphere%Aerosol(i)%Type == -1) CYCLE
        Aerosol_Type(j)         = Atmosphere%Aerosol(i)%Type
        Effective_Radius(j,:)   = Atmosphere%Aerosol(i)%Effective_Radius
        Effective_Variance(j,:) = Atmosphere%Aerosol(i)%Effective_Variance
        Concentration(j,:)      = Atmosphere%Aerosol(i)%Concentration
        j = j + 1
    END DO

    ! Update Atmosphere data sturcture
    Atmosphere%n_Aerosols = n_aerosol_valid
    DO j = 1, n_aerosol_valid
      Atmosphere%Aerosol(j)%Type             = Aerosol_Type(j)
      Atmosphere%Aerosol(j)%Effective_Radius = Effective_Variance(j,:)
      Atmosphere%Aerosol(j)%Concentration    = Concentration(j,:)
    END DO

  END SUBROUTINE CRTM_Atmosphere_Aerosol_Bypass
```

Later realize that this does not work because the output aerosol index will be rearranged, further causing problem in DA applications.

**Comment by chengdang** [comment]
So the only option is to skip the aerosol-related LUT interpolation, if the type index is -1. The following conditional statement is added in subroutines Get_Aerosol_Opt, Get_Aerosol_Opt_TL, Get_Aerosol_Opt_AD, (https://github.com/JCSDA/CRTMv3/blob/feature/cd_aero_bypass/src/AtmScatter/CRTM_AerosolScatter.f90)

`IF ( Aerosol_Type == -1 ) RETURN  ! Bypass aerosol`



Type = -1 is set as an valid in `Function CRTM_Aerosol_IsValid`
https://github.com/JCSDA/CRTMv3/blob/1bdae5f0c63680675d50a0a5da7818de53eecfe1/src/Atmosphere/Aerosol/CRTM_Aerosol_Define.f90#L507

**Comment by chengdang** [comment]
A unit test has also been added to test if this implementation works in CRTM standalone mode:
https://github.com/JCSDA/CRTMv3/blob/feature/cd_aero_bypass/test/mains/unit/Unit_Test/test_Aerosol_Bypass.f90

With ICASE = 1, CRTM will read the first two types of aerosols only. (Load_Atm_Data.inc)
With ICASE = 2, CRTM will read all three types of aerosols, but the index of the third one is -1, so it is equivalent to 2 types of aerosols. (Load_Atm_Data_Bypass_Aerosol.inc)
The RT solutions computed with these two cases are indentical.

I have not tested this with the TL/AD modules nor with a large dataset.

**Comment by chengdang** [comment]
Testing by @jeromebarre reveals minor difference of AOD simulation, since nitrate aerosols are bypassed, the AOD should be identical.

Using CRTM LUT with GEFS option
```
ObsAodCRTM variables: 20 variables: air_temperature, humidity_mixing_ratio, relative_humidity, air_pressure, air_pressure_levels, mass_fraction_of_sulfate_in_air, mass_fraction_of_hydrophobic_black_carbon_in_air, mass_fraction_of_hydrophilic_black_carbon_in_air, mass_fraction_of_hydrophobic_organic_carbon_in_air, mass_fraction_of_hydrophilic_organic_carbon_in_air, mass_fraction_of_dust001_in_air, mass_fraction_of_dust002_in_air, mass_fraction_of_dust003_in_air, mass_fraction_of_dust004_in_air, mass_fraction_of_dust005_in_air, mass_fraction_of_sea_salt001_in_air, mass_fraction_of_sea_salt002_in_air, mass_fraction_of_sea_salt003_in_air, mass_fraction_of_sea_salt004_in_air, mass_fraction_of_sea_salt005_in_air
124: ObsAodCRTM channels: [4]
124: No bias-correction is performed for this ObsSpace.
124: Vector difference between reference and computed: Aod nobs= 100 Min=-0.000573758, Max=7.19802e-06, RMS=9.90383e-05
```

Using CRTM LUT with UFS option
```
124: ObsAodCRTM variables: 23 variables: air_temperature, humidity_mixing_ratio, relative_humidity, air_pressure, air_pressure_levels, mass_fraction_of_sulfate_in_air, mass_fraction_of_hydrophobic_black_carbon_in_air, mass_fraction_of_hydrophilic_black_carbon_in_air, mass_fraction_of_hydrophobic_organic_carbon_in_air, mass_fraction_of_hydrophilic_organic_carbon_in_air, mass_fraction_of_dust001_in_air, mass_fraction_of_dust002_in_air, mass_fraction_of_dust003_in_air, mass_fraction_of_dust004_in_air, mass_fraction_of_dust005_in_air, mass_fraction_of_sea_salt001_in_air, mass_fraction_of_sea_salt002_in_air, mass_fraction_of_sea_salt003_in_air, mass_fraction_of_sea_salt004_in_air, mass_fraction_of_sea_salt005_in_air, mass_fraction_of_nitrate001_in_air, mass_fraction_of_nitrate002_in_air, mass_fraction_of_nitrate003_in_air
124: ObsAodCRTM channels: [4]
124: No bias-correction is performed for this ObsSpace.
124: Vector difference between reference and computed: Aod nobs= 100 Min=-0.000573773, Max=6.4792e-06, RMS=9.90608e-05
```

I agree the AOD should be identical. Went over the code again, that there is no obvious error. Will continue working on this.

**Comment by jeromebarre** [comment]
with
```
       AerosolCoeff_File: AerosolCoeff.nc4
       AerosolCoeff_Format: netCDF
       AerosolOption: aerosols_gocart_default
```

 ```
375: ObsAodCRTMTLAD variables: 14 variables: mass_fraction_of_sulfate_in_air, mass_fraction_of_hydrophobic_black_carbon_in_air, mass_fraction_of_hydrophilic_black_carbon_in_air, mass_fraction_of_hydrophobic_organic_carbon_in_air, mass_fraction_of_hydrophilic_organic_carbon_in_air, mass_fraction_of_dust001_in_air, mass_fraction_of_dust002_in_air, mass_fraction_of_dust003_in_air, mass_fraction_of_dust004_in_air, mass_fraction_of_dust005_in_air, mass_fraction_of_sea_salt001_in_air, mass_fraction_of_sea_salt002_in_air, mass_fraction_of_sea_salt003_in_air, mass_fraction_of_sea_salt004_in_air
 375: ObsAodCRTMTLAD channels: [4]
 375: No bias-correction is performed for this ObsSpace.
 375: No bias-correction is performed for this ObsSpace.
 375: Iter:0 ||(h(x+alpha*dx)-h(x)-h'*(alpha*dx))||=1.23251e-06
 375: Completed case 2: interface/LinearObsOperator/testTangentLinear
```

the TL test runs ok

but with
```
AerosolCoeff_File: AerosolCoeff.nc4
       AerosolCoeff_Format: netCDF
       AerosolOption: aerosols_gocart_ufs
```

 ```
 ObsAodCRTMTLAD variables: 18 variables: mass_fraction_of_sulfate_in_air, mass_fraction_of_hydrophobic_black_carbon_in_air, mass_fraction_of_hydrophilic_black_carbon_in_air, mass_fraction_of_hydrophobic_organic_carbon_in_air, mass_fraction_of_hydrophilic_organic_carbon_in_air, mass_fraction_of_dust001_in_air, mass_fraction_of_dust002_in_air, mass_fraction_of_dust003_in_air, mass_fraction_of_dust004_in_air, mass_fraction_of_dust005_in_air, mass_fraction_of_sea_salt001_in_air, mass_fraction_of_sea_salt002_in_air, mass_fraction_of_sea_salt003_in_air, mass_fraction_of_sea_salt004_in_air, mass_fraction_of_sea_salt005_in_air, mass_fraction_of_nitrate001_in_air, mass_fraction_of_nitrate002_in_air, mass_fraction_of_nitrate003_in_air
 375: ObsAodCRTMTLAD channels: [4]
 375: No bias-correction is performed for this ObsSpace.
 375: No bias-correction is performed for this ObsSpace.
 375: Iter:0 ||(h(x+alpha*dx)-h(x)-h'*(alpha*dx))||=inf
 375: Test "interface/LinearObsOperator/testTangentLinear" failed: Condition failed: y2.rms() < tol @  (/work2/noaa/jcsda/barre/jedi-release/jedi-bundle/oops/src/test/interface/LinearObsOperator.h +447 testTangentLinear)
 375: Completed case 2: interface/LinearObsOperator/testTangentLinear
```

The TL test returns infinity. It could indicate a division by 0 somewhere.

**Comment by chengdang** [comment]
With TL/AD output, we also have to by pass calculations with aerosol with index -1. Depending on how the atm_TL is used in JEDI, it would not be surprising if there is a dividing by zero somewhere.
I'll write a CRTM unit test for TL/AD calculations first.

**Comment by jeromebarre** [comment]
I think this is very dangerous to do this: https://github.com/JCSDA/CRTMv3/blob/1bdae5f0c63680675d50a0a5da7818de53eecfe1/src/AtmScatter/CRTM_AerosolScatter.f90#L799
as it left potentially a lot of unallocated stuff. 
I think in this condition we we need to make sure the optical properties returned are transparent i.e. resulting extinctions, albedos and radii are set to 0. If possible and not too much work and code complications.

**Comment by jeromebarre** [comment]
same goes in the TL and AD. will come up with a commit in the branch and let you know.

**Comment by chengdang** [comment]
@jeromebarre's latest commit fixed the issues reported earlier, see commit https://github.com/JCSDA/CRTMv3/pull/115/commits/9a8821fa4d91c2c3b0f63efc131ce1f5720abe27. Thank you!

**Comment by chengdang** [comment]
With commit https://github.com/JCSDA/CRTMv3/commit/b964cb34ee35f6c12334b8a961ef1371229811e6,

1. Added TL, adjoint, and k-matrix tests for aerosol bypass. Note the atm_K comparisons in the adjoint and k-matrix tests are removed because they will fail for sure with the different number of aerosols in two cases.
2. Replaced index -1 with new parameter AerosolCoeff_BYPASS_AEROSOL in all related codes.
3. Fixed error in CRTM_Atmosphere_Inspect() when an aerosol type is set as AerosolCoeff_BYPASS_AEROSOL. https://github.com/JCSDA/CRTMv3/blob/b964cb34ee35f6c12334b8a961ef1371229811e6/src/Coefficients/AerosolCoeff/AerosolCoeff_Define.f90#L846
https://github.com/JCSDA/CRTMv3/blob/b964cb34ee35f6c12334b8a961ef1371229811e6/src/Atmosphere/Aerosol/CRTM_Aerosol_Define.f90#L150
Without this update, CRTM will pause after calling `CRTM_Atmosphere_Inspect()`, `CRTM_Aerosol_CategoryName()`, and report a segmentation fault.

**Comment by chengdang** [comment]
closed with PR #115
